proposals/lip-0054.md

```diff
@@ -301,104 +303,141 @@ messageRecoveryParams = {
-    if ccm.receivingChainID == ownChainAccount.ID:
-        if (ccm.moduleID, ccm.crossChainCommandID) is supported:
-            call the logic associated with (ccm.moduleID, ccm.crossChainCommandID) on ccm
-    elif ownChainAccount.ID == MAINCHAIN_ID:
-        # Only send messages to active and live chains
-        if (chainAccount(ccm.receivingChainID) exists
-            and chainAccount(ccm.receivingChainID).status == CHAIN_ACTIVE
+def execute(trs: Transaction) -> None:
+    senderAddress = SHA-256(trs.senderPublicKey)[:ADDRESS_LENGTH]
+    # Set CCM status to recovered and assign fee to trs sender
+    crossChainMessages = [decode(crossChainMessageSchema, ccmBytes) for ccmBytes in trs.params.crossChainMessages]
+    updatedCCMs = []
+    for ccm in crossChainMessages:
+        token.assignMessageFee(senderAddress, ccm)
+        ccm.status = CCM_STATUS_RECOVERED
+        ccm.fee = 0
+        updatedCCMs.append(encode(crossChainMessageSchema, ccm))
+
+    # Update sidechain outbox root
+    proof = { 
+        "size": terminatedOutboxAccount(trs.params.chainID).outboxSize,
+        "idxs": trs.params.idxs,
+        "siblingHashes": trs.params.siblingHashes
+    }
+
+    terminatedOutboxAccount(trs.params.chainID).outboxRoot = RMTCalculateRoot([SHA-256(ccmData) for CCMData in updatedCCMs], proof)
+
+    # Process recovery.
+    for ccm in crossChainMessages:
+        ccmID = SHA-256(encode(crossChainMessageSchema, ccm))
+        # On sidechain, this case is always triggered and the message logic should be supported
+        if ccm.receivingChainID == ownChainAccount.ID and (ccm.moduleID, ccm.crossChainCommandID) is supported:
+            # Create a state snapshot.
+            snapshot = snapshot of the current state
+            try: 
+                # Emit processing event. If an error occurs, this event emission is also undone.
+                emitEvent(
+                    moduleID = MODULE_ID_INTEROPERABILITY,
+                    typeID = TYPE_ID_CCM_PROCESSED,
+                    data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_COMMAND_RECOVERY_APPLIED},
+                    topics = [ccm.sendingChainID, ccm.receivingChainID]
+                )
+
+                # Call the beforeCrossChainCommandExecution functions from other modules
+                for each module mdl for which beforeCrossChainCommandExecution exists:
+                    mdl.beforeCrossChainCommandExecution(trs, ccm)
+
+                # Execute the cross-chain command
+                crossChainCommand = cross-chain command associated with (ccm.moduleID, ccm.crossChainCommandID)
+                crossChainCommand.execute(ccm)
+
+                # Call the afterCrossChainCommandExecution functions from other modules
+                for each module mdl for which beforeCrossChainCommandExecution exists:
+                    mdl.afterCrossChainCommandExecution(trs, ccm)
+            except:
+                revert state to snapshot 
```

Comment:
Remove this line and raise an Exception after emitting persistent event.

Reply:
I think we should indeed revert on the state of snapshot here..